File: 454.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int fourSumCount(vector<int>& nums1, vector<int>& nums2, vector<int>& nums3, vector<int>& nums4) {
        unordered_map<int, int> ma;
        unordered_map<int, int> md;
        for (int a : nums1) {
            for (int b : nums2) {
                ma[a + b]++;
            }
        }
        int cnt = 0;
        for (int c : nums3) {
            for (int d : nums4) {
                if (ma.find(-(c + d)) != ma.end())
                    cnt += ma.find(-(c + d))->second;
            }
        }
        return cnt;
    }
};

File: 454.go
package leetcode

import "sort"

func fourSumCount(nums1 []int, nums2 []int, nums3 []int, nums4 []int) int {
	sort.Ints(nums1)
	sort.Ints(nums2)
	sort.Ints(nums3)
	umap := make(map[int]int)
	for i := 0; i < len(nums4); i++ {
		for j := 0; j < len(nums3); j++ {
			umap[nums4[i] + nums3[j]]++
		}
	}
	cnt := 0
	for i := 0; i < len(nums1); i++ {
		for j := 0; j < len(nums2); j++ {
			cnt += umap[-(nums1[i] + nums2[j])]
		}
	}
	return cnt
}